File: backend/app/schemas/social_spark.py
"""
LAYERS - Social Spark Schemas
=============================================
Pydantic models for the boost / wave / synchronicity endpoints.
"""

from datetime import datetime
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, Field


# ============================================================
# SIGNAL BOOST  📡
# ============================================================

class BoostCreateRequest(BaseModel):
    """POST /spark/artifacts/{id}/boost — no body fields needed beyond path."""
    # Reserved for future tuning (e.g. custom radius). Empty for now.
    pass


class BoostResponse(BaseModel):
    id: UUID
    artifact_id: UUID
    booster_id: UUID
    boost_radius_meters: int
    created_at: datetime
    expires_at: datetime

    model_config = {"from_attributes": True}


class BoostQuotaResponse(BaseModel):
    """How many boosts the user has left today."""
    used_today: int
    daily_limit: int
    remaining: int


class BoostedArtifactItem(BaseModel):
    """A boosted artifact surfaced on the map beyond the normal radius."""
    artifact_id: UUID
    latitude: float
    longitude: float
    distance_meters: float
    boost_expires_at: datetime


class BoostedNearbyResponse(BaseModel):
    items: List[BoostedArtifactItem]


# ============================================================
# WAVE  👋
# ============================================================

class WaveCreateRequest(BaseModel):
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)


class WaveCreateResponse(BaseModel):
    """
    Result of dropping a wave. We never reveal who waved — only how many
    other people waved nearby in the recent window (so it feels reciprocal).
    """
    wave_id: UUID
    expires_at: datetime
    others_waving_nearby: int
    waved_back: bool  # True if someone else had a live wave here when you waved


class WaveNearbyResponse(BaseModel):
    """GET /spark/waves/nearby — fully anonymous."""
    count: int  # Active waves within the radius (excluding your own)
    radius_meters: int


# ============================================================
# SYNCHRONICITY  ✨
# ============================================================

class DiscoverRequest(BaseModel):
    """
    POST /spark/artifacts/{id}/discover — call when an artifact unlocks
    on the client (within 50m, content revealed). Idempotent per user.
    """
    latitude: Optional[float] = Field(None, ge=-90, le=90)
    longitude: Optional[float] = Field(None, ge=-180, le=180)


class SynchronicityMatch(BaseModel):
    """A single sync — anonymized: we never name the other person."""
    event_id: UUID
    artifact_id: UUID
    created_at: datetime


class DiscoverResponse(BaseModel):
    """
    Tells the client whether this discovery sparked a synchronicity.
    `is_new_discovery` is False on repeat calls (idempotent no-op).
    """
    is_new_discovery: bool
    synchronicity: Optional[SynchronicityMatch] = None
    message: Optional[str] = None  # e.g. "Someone else felt this too ✨"


class SynchronicityListItem(BaseModel):
    event_id: UUID
    artifact_id: UUID
    created_at: datetime

    model_config = {"from_attributes": True}


class SynchronicityListResponse(BaseModel):
    items: List[SynchronicityListItem]
    total: int

File: mobile/src/types/social_spark.ts
/**
 * LAYERS — Social Spark Types (Week 6 Day 4)
 * ============================================
 * Mirrors backend/app/schemas/social_spark.py.
 *
 *   📡 Signal Boost   — amplify an artifact
 *   👋 Anonymous Wave — ephemeral "I'm here too"
 *   ✨ Synchronicity  — two strangers, same artifact, 30 min
 */

// ============================================================
// 📡 BOOST
// ============================================================

export interface BoostResponse {
  id: string;
  artifact_id: string;
  booster_id: string;
  boost_radius_meters: number;
  created_at: string;
  expires_at: string;
}

export interface BoostQuota {
  used_today: number;
  daily_limit: number;
  remaining: number;
}

export interface BoostedArtifactItem {
  artifact_id: string;
  latitude: number;
  longitude: number;
  distance_meters: number;
  boost_expires_at: string;
}

export interface BoostedNearbyResponse {
  items: BoostedArtifactItem[];
}

// ============================================================
// 👋 WAVE
// ============================================================

export interface WaveCreateResponse {
  wave_id: string;
  expires_at: string;
  others_waving_nearby: number;
  waved_back: boolean;
}

export interface WaveNearbyResponse {
  count: number;
  radius_meters: number;
}

// ============================================================
// ✨ SYNCHRONICITY
// ============================================================

export interface SynchronicityMatch {
  event_id: string;
  artifact_id: string;
  created_at: string;
}

export interface DiscoverResponse {
  is_new_discovery: boolean;
  synchronicity: SynchronicityMatch | null;
  message: string | null;
}

export interface SynchronicityListItem {
  event_id: string;
  artifact_id: string;
  created_at: string;
}

export interface SynchronicityListResponse {
  items: SynchronicityListItem[];
  total: number;
}
